chore: improved benchmark reporting and gating

MarkMcCulloh opened the pull request:
Fixes #428

- In CI, the benchmarking will fail if any of the benches are slower than main by a percentage threshold
  - This threshold is currently set to a stupid number (50%). I'd like to ease it down to find the sweet spot over time
- Benchmarks should hopefully be more reliable
  - Running more iterations
  - Added a small sleep between each run, noticed that process handled leaked for a short time in-between cycles
  - (Note) None of the above effects the final runtime of CI since it still runs faster than hangar
- Added a comment to the PR with bench results and comparisons to main
  - This only happens for PRs from this repo (for now)
- Locally, compare to your last run
- Added a bench to get the version from the CLI
  - kinda works as a nice baseline


After this merges, I think it would be valuable to use the `wing -V` result as a baseline to determine if the threshold is being breached. Right now the variation of runs is still pretty bad. The last option left would be to get dedicated machines for benchmarks.

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

MarkMcCulloh commented on tools/hangar/src/benchmarking/table_report.ts:
This file was moved from scripts/benchmark_json_to_table.mjs.

Other than switching it to typescript, it has barely changed.

staycoolcall911 commented on tools/hangar/src/benchmarking/compare.ts:
```diff
@@ -0,0 +1,135 @@
+    if (num === undefined || isNaN(num)) {
+      return "...";
+    }
+    return num.toFixed(2) + unit;
```
It applies to all the measurements we finally print to the table: I think we can keep it at ms, no reason for 1/100 of ms precision.

MarkMcCulloh replied:
Ok I changed ms to be rounded, but I kept 1 decimal of precision of the % (to prevent it from saying 0% change)

staycoolcall911 reviewed (commented):
We've been expecting this! Great!!

staycoolcall911 reviewed (approved):
LGTM

Chriscbr commented on .github/workflows/build.yml:
```diff
@@ -25,6 +25,7 @@ env:
   CARGO_TERM_COLOR: always
   TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
   TURBO_TEAM: ${{ vars.TURBO_TEAM }}
+  SAME_REPO_PR: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository }}
```
```suggestion
  IS_SAME_REPO_PR: ${{ github.event_name == 'pull_request' && github.event.pull_request.head.repo.full_name == github.repository }}
```

Chriscbr reviewed (approved):
Nice work! A few non-blocking comments:

- Is showing all of the samples a bit verbose? (maybe we can just link to the build logs if someone wants to see the exact numbers)
- `totalTime` feels redundant (it's just `mean` * # of samples). Maybe we could show `variance` / `stdDev` instead
- Can we include plus/minus signs (+/-) in the `Change` column to signal whether the changes were increases or decreases?  Bonus points we can highlight them as green or red for easier-at-a-glance readability
- Super nit: move benchmark Wing files to `examples/tests/benchmarks` to make them easier to discover. (If benchmarking errors in CI, that should fail the build, so I'm not sure if they also need to be ran by the hangar)
- Can we add a section to the contributing guide about how to run the benchmarks locally (if it's possible)?

MarkMcCulloh commented:
> Super nit: move benchmark Wing files to examples/tests/benchmarks to make them easier to discover. (If benchmarking errors in CI, that should fail the build, so I'm not sure if they also need to be ran by the hangar)

Addressed all of the feedback except this. Having the benchmarks as a subset of `valid` is useful because it's shared node_modules with the rest of them. Also, I can see a usecase to have invalid benchmarks at some point.
